### PR TITLE
Support todatetimeoffset function

### PR DESCRIPTION
**Pull request by hash-16**
### Description

This function returns a datetimeoffset value that is translated from a datetime2 expression.

Example:-

```
SELECT TODATETIMEOFFSET(SYSDATETIME(), -120)
2023-08-03 10:26:35.5053494 -02:00

Select SYSDATETIME()
2023-08-03 10:27:07.3436982
```

Signed-off-by: Ashish Prasad [pashisht@amazon.com](mailto:pashisht@amazon.com)

### Issues Resolved

BABEL-740

### Test Scenarios Covered ###

NOTE :-  There are issues with cast to datetimeoffset ,datetime2 datatype , filed a JIRA (BABEL-4328) , so there are some testcase which fail during cast to datetimeoffset which will also fail for the todatetimeoffset function.

* **Use case based -** Added

* **Boundary conditions -** Added

* **Arbitrary inputs -** Added

* **Negative test cases -** Added

* **Minor version upgrade tests -** Added

* **Major version upgrade tests -** Added

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** NA

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/python/expected/upgrade_validation/expected_dependency.out [hunk 1]
```diff
@@ -632,6 +632,7 @@ Function sys.timestamptz2smalldatetime(timestamp with time zone)
 Function sys.tinyint_larger(sys.tinyint,sys.tinyint)
 Function sys.tinyint_smaller(sys.tinyint,sys.tinyint)
 Function sys.tinyint_sqlvariant(sys.tinyint)
+Function sys.todatetimeoffset(text,text)
```
Comment (shalinilohia50) [inline test/python/expected/upgrade_validation/expected_dependency.out]: Add dependency test for this function call.

Comment (hash-16) [reply test/python/expected/upgrade_validation/expected_dependency.out]: Added

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 2]
```diff
@@ -353,6 +353,119 @@ LANGUAGE plpgsql
...
+    v_hr := str_hr::INTEGER;
+    v_mi := str_mi ::INTEGER;
+    exception
+        WHEN others THEN
+            RAISE USING MESSAGE := 'The timezone provided to builtin function todatetimeoffset is invalid.';
+    END;
+
+    v_hr := v_hr * sign_flag;
+    
+    if v_hr > 14 or v_hr < -14 or (v_hr = 14 and v_mi > 0) or (v_hr = -14 and v_mi > 0) THEN
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: why is this condition diff between switch offset vs todatetimeoffset?

File: test/JDBC/expected/todatetimeoffset.out [hunk 3]
```diff
@@ -0,0 +1,204 @@
+Select todatetimeoffset('0000-04-22 16:23:51.766890','+12:00')
```
Comment (surendravishnoi) [inline test/JDBC/expected/todatetimeoffset.out:1]: Please add test case for each exception condition. Test cases for boundary values (min/max/0), invalid values, data types and formats. Also, add test cases where values are coming from a table, input is expr, output is fed to expr etc.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 2 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Updated

File: test/JDBC/expected/todatetimeoffset.out (hunk 3 above)
Comment (hash-16) [reply test/JDBC/expected/todatetimeoffset.out:1]: Added

File: test/JDBC/expected/todatetimeoffset.out [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+
+
+Select todatetimeoffset('2000-04-22 16:23:5d.766890','+12:00')
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Conversion failed when converting date and/or time from character string.)~~
+
+
+Select todatetimeoffset('2000-04-22 16:23:51.7668c0','+12:00')
```
Comment (surendravishnoi) [inline test/JDBC/expected/todatetimeoffset.out]: is this expected, not to fail for 7668c0?

File: test/JDBC/expected/todatetimeoffset.out [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+
+
+Select todatetimeoffset('9934-11-30 22:52:59.59', '-14:01')
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: The timezone provided to builtin function todatetimeoffset is invalid.)~~
+
+
+Select todatetimeoffset(CAST('1900-05-06 13:59:29.998 -8:00' AS datetime2), '+12:00')
```
Comment (surendravishnoi) [inline test/JDBC/expected/todatetimeoffset.out:259]: what happened to -8:00?

File: test/JDBC/expected/todatetimeoffset.out [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+Go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Conversion failed when converting date and/or time from character string.)~~
+
+
+DROP TABLE IF EXISTS tem
+GO
+Create table tem(a datetimeoffset)
+insert into tem (a) values(todatetimeoffset('2000-04-22 12:23:51.766890',120))
```
Comment (surendravishnoi) [inline test/JDBC/expected/todatetimeoffset.out:320]: let also do todatetimeoffset on column a as part of select

File: test/JDBC/expected/todatetimeoffset.out (hunk 4 above)
Comment (hash-16) [reply test/JDBC/expected/todatetimeoffset.out]: Yeah this is expected and this is due to issue in the cast for datetime2 and datetimeoffset function.
```
1> Select cast('2000-04-22 16:23:51.7668c0' as datetime2)
2> go
datetime2                             
--------------------------------------
           2000-04-22 16:23:51.7668000

(1 rows affected)
1> Select cast('2000-04-22 16:23:51.7668c0' as datetimeoffset)
2> go
datetimeoffset                               
---------------------------------------------
           2000-04-22 16:23:51.7668000 +00:00
```

File: test/JDBC/expected/todatetimeoffset.out (hunk 5 above)
Comment (hash-16) [reply test/JDBC/expected/todatetimeoffset.out:259]: That will be replaced by the new timezoneoffset as expected .
```

1> Select todatetimeoffset(CAST('1900-05-06 13:59:29.998 -8:00' AS datetime2), '+12:00')
2> go
                                             
---------------------------------------------
           1900-05-06 13:59:29.9980000 +12:00
```

File: test/JDBC/expected/todatetimeoffset.out (hunk 4 above)
Comment (surendravishnoi) [reply test/JDBC/expected/todatetimeoffset.out]: you mean this is not expected but happens due to cast issue for dateime2/datetimeoffset. If that is the case, please add a comment with corresponding Jira.

File: test/JDBC/expected/todatetimeoffset.out (hunk 6 above)
Comment (hash-16) [reply test/JDBC/expected/todatetimeoffset.out:320]: Added

File: test/JDBC/expected/todatetimeoffset.out (hunk 4 above)
Comment (hash-16) [reply test/JDBC/expected/todatetimeoffset.out]: Addressed.

**Review by surendravishnoi — DISMISSED** [review]

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 7]
```diff
@@ -353,6 +353,159 @@ LANGUAGE plpgsql
 IMMUTABLE
 RETURNS NULL ON NULL INPUT;
 
+
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: Nit: Extra lines

**Review by shalinilohia50 — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 7 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Addressed

**Review by surendravishnoi — APPROVED** [review]

**Pull request merged**